01-data-structures/03-linked-lists/linked_list.rb

```diff
@@ -4,27 +4,120 @@ class LinkedList
   attr_accessor :head
   attr_accessor :tail
 
+  def initialize
+    @head = nil
+    @tail = nil
+  end
+
+  def find_index(number)
+    current = @head
+    count = 1
+    until count == number
+      current = current.next
+      count += 1
+    end
+    current
+  end
+
+  def delete_index(number)
+    current = @head
+    count = 1
+    until count == number
+      current = current.next
+      count += 1
+    end
+    self.delete(current)
+  end
+
   # This method creates a new `Node` using `data`, and inserts it at the end of the list.
   def add_to_tail(node)
+    if @tail.nil?
+      @head = node
+    else
+      @tail.next = node
+    end
+    @tail = node
   end
 
   # This method removes the last node in the lists and must keep the rest of the list intact.
   def remove_tail
+    return nil if @head == nil
+
+    node = @head
+    popped = @tail
+    if @head == @tail
+      @head = nil
+      @tail = nil
+    else
+      until node.next == @tail
+        node = node.next
+      end
+      @tail = node
+      @tail.next = nil
+    end
   end
 
   # This method prints out a representation of the list.
   def print
+    return nil if @head.nil?
+    print_string = ""
+    current = @head
+    print_string.concat(current.data + "\n")
+    until current.next.nil?
+      current = current.next
+      print_string.concat(current.data + "\n")
+    end
+    puts print_string
   end
 
   # This method removes `node` from the list and must keep the rest of the list intact.
   def delete(node)
+    removed = nil
```

Comment:
I think this could be refactored to be cleaner

Reply:
But I like it the way it is!
Can't we keep it :) pleeeeease?
The delete method is complicated but it makes sense, 
if head .... if tail .... if in the middle .................................. easy peasy!
I don't know how else to right it more succinctly

Reply:
It's very difficult to read right now.

Reply:
I would rewrite in English what you're trying to accomplish and see if there's a way you could do it in fewer lines